db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -4096,6 +4096,9 @@ Status DBImpl::WaitForCompact(
       return s;
     }
   }
+  if (wait_for_compact_options.close_db) {
```

Comment:
I wonder for when `close_db` is true, do we not only need to cancel scheduling any of the tasks that have been registered to it, but also prohibit the task scheduler from taking any future task registrations? Something like adding a `PeriodicTaskScheduler::Stop()` API that will mark the scheduler to not take any further `PeriodicTaskScheduler::Register` calls.

Reply:
BTW, why does `CancelPeriodicTaskSchedulers` be called here since it will be invoke in `Close()` anyway when `close_db` is true?

Reply:
For the first comment, I think that's a valid point. I'm wondering where we register more background jobs outside of db open, though. If we register new periodic tasks only during open, then I think it may not be necessary today, but it could be a safer approach in case there's a future change where we register new periodic tasks outside of db open.

For the second point, between the very short time unlocking db mutex and the cancelling the periodic tasks during Close(), it's still possible that we may have background jobs scheduled. So we decided to cancel the periodic tasks before waiting although this is called again during Close().

Reply:
I don't quite get this, I think periodic task scheduler won't be scheduling the type of background work the wait loop below is concerned about though (https://github.com/facebook/rocksdb/blob/main/db/periodic_task_scheduler.h#L20-L26).

Reply:
>I think periodic task scheduler won't be scheduling the type of background work the wait loop below is concerned about though

Thanks for pointing this out. I didn't look deep enough. I basically assumed that the periodic tasks would include a task that triggers the periodic compaction. I read the code again and I now see that my assumption was wrong. 

IIUC, periodic compaction is not actually something that's always scheduled periodically. When compaction score gets calculated, files get marked for periodic compaction based on the file's last modification time and the db option, then when compaction picker gets invoked to pick compaction, it checks for files that are marked for periodic compaction and pick them to be compacted instead of other priorities.

So I guess the cancelling the periodic tasks before waiting isn't really necessary here anymore. I'm still trying to think of any scenario that may add more background work after waiting, during the short time of mutex unlock and `shutting_down_` is set true, though.

Reply:
If we are concerned about after the waiting is done, no more background work added after we unlock the mutex. One idea is to have a `atomic<bool> accepting_background_job_` to guard `flush_queue_` and `compaction_queue_`, and set the flag to false before unlock the mutex.

We need to make accompanying adjustment to `DBImpl::SchedulePendingCompaction` and `DBImpl::SchedulePendingFlush` to check this guarding flag and return immediately when it's false.

Another refactoring needed is that the wait loop below checks counters `unscheduled_compactions_` and `unscheduled_flushes_` which are supposed to be a one to one mapping for the number of jobs in `compaction_queue_` and `flush_queue_` respectively. `unscheduled_flushes_` is exclusively incremented only in `DBImpl::SchedulePendingFlush`, but some caller is using `DBImpl::SchedulePendingCompaction` which adds to `unscheduled_compactions_` automatically, some caller is using `DBImpl::AddToCompactionQueue` and do its own incrementation, some refactoring is needed to make sure the counter shows the correct number of jobs after this guarding flag change.

This helps with not getting more background work after waiting is done, but I don't know if it helps with guarantee that "no background work would happen upon re-opening the db" as discussed in the other thread. Is that a requirement of this API too?

Reply:
The method names are a bit confusing to me `SchedulePendingCompaction()` and `SchedulePendingFlush()` in fact do not actually "schedule" the jobs in background. They add the jobs to the queue and jobs are still considered "unscheduled" and `MaybeScheduleFlushOrCompaction()` actually calls `env_->Schedule()` to schedule the queued jobs to run in the background.

Thanks for the suggestion. I wanted to avoid introducing new flags which could potentially have larger impact than maybe necessary. I believe that setting `shutting_down_ ` true before unlocking the mutex should be sufficient to prevent new jobs to be scheduled in the background because `MaybeScheduleFlushOrCompaction()` will early return out. Without implementing your suggestion, however, it won't prevent new jobs to be added to the queue ( a.k.a. "unscheduled"). I believe that this is an acceptable behavior.

To answer your last question, my understanding is that we will make the _best effort_ to finish all compaction (and flushing if flush=true) needed before close, but is that a hard requirement? that's a good question. I think not, but I'd love to hear what @pdillinger  thinks.

Reply:
Actually, after some thoughts, setting `shutting_down_` prior to `Close()` may cause some problems like,
if `Close()` is aborted because there's still snapshot to be released. `MaybeReleaseTimestampedSnapshotsAndCheck()` also needs db lock, so I cannot call this before releasing.

I'm going to introduce new booleans as you suggested, but with `false` by default. (i.e. `reject_new_background_jobs_`) to prevent new background jobs to be added after waiting. Thanks again for the review and suggestion. Will let you know when it's ready for another round of review.